src/common/column_matrix.h

```diff
@@ -12,52 +12,67 @@
 #include <vector>
 #include "hist_util.h"
 
-
 namespace xgboost {
 namespace common {
 
-
+class ColumnMatrix;
 /*! \brief column type */
 enum ColumnType {
   kDenseColumn,
   kSparseColumn
 };
 
 /*! \brief a column storage, to be used with ApplySplit. Note that each
-    bin id is stored as index[i] + index_base. */
+    bin id is stored as index[i] + index_base.
+    Different types of column index for each column allow
+    to reduce the memory usage. */
+template <typename BinIdxType>
 class Column {
  public:
-  Column(ColumnType type, const uint32_t* index, uint32_t index_base,
-         const size_t* row_ind, size_t len)
+  Column(ColumnType type, common::Span<const BinIdxType> index,
+         uint32_t index_base, const size_t* row_ind,
+         const std::vector<bool>::const_iterator missing_flags)
       : type_(type),
         index_(index),
         index_base_(index_base),
         row_ind_(row_ind),
-        len_(len) {}
-  size_t Size() const { return len_; }
-  uint32_t GetGlobalBinIdx(size_t idx) const { return index_base_ + index_[idx]; }
-  uint32_t GetFeatureBinIdx(size_t idx) const { return index_[idx]; }
-  common::Span<const uint32_t> GetFeatureBinIdxPtr() const { return { index_, len_ }; }
-  // column.GetFeatureBinIdx(idx) + column.GetBaseIdx(idx) ==
-  // column.GetGlobalBinIdx(idx)
+        missing_flags_(missing_flags) {}
+
+  // column.GetGlobalBinIdx(idx) == column.GetFeatureBinIdx(idx) + column.GetBaseIdx(idx)
```

Comment:
I agree, it can be confusing, but it's not a code part. This line is just about additional explanation how methods below are related to each other. 
If it's obvious and comment is redundant I can deleted it.